### PR TITLE
Fix ArgumentOutOfRangeException in Option strike filtering

### LINKED ISSUE #1875: The Option SetFilter() method throws "out of range" error for "GOOG"
For GOOG options from 04/06/2016 to 04/08/2016, the SetFilter(minStrike, maxStrike) method with negative maxStrike input would throw error `Runtime Error: Index was out of range. Must be non-negative and less than the size of the collection`. The error occurs at `2016-04-07 19:57:00` and `2016-04-08 09:30:00` which is not trading hours for options.
#### Expected Behavior

SetFilter(minStrike, maxStrike) method should allow negative maxStrike input. 
#### Actual Behavior
Negative maxStrike input in SetFilter() would throw errors of index out of range.
#### Reproducing the Problem

``` csharp
namespace QuantConnect.Algorithm.CSharp
{
    public class BasicTemplateOptionsAlgorithm : QCAlgorithm
    {
        public override void Initialize()
        {
            SetStartDate(2016, 4, 6);
            SetEndDate(2016, 4, 8);
            SetCash(100000);
            var equity = AddEquity("GOOG", Resolution.Minute);
            var option = AddOption("GOOG");
            option.SetFilter(-5, -1, TimeSpan.FromDays(14), TimeSpan.FromDays(50));
        }
        public override void OnData(Slice slice)
        {
        }
    }
}
```

#### Checklist

- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

- [x] I have provided detailed steps to reproduce the issue

### PR DESCRIPTION
**Pull request by StefanoRaggi**
#### Description
The option strike range filtering in `OptionFilterUniverse.Strikes` has been updated to support edge cases such as the current underlying price being outside the strike range.

#### Related Issue
Fixes #1875

#### Motivation and Context
The method was throwing `ArgumentOutOfRangeException` in specific cases.

#### Requires Documentation Change
No.

#### How Has This Been Tested?
Tested manually with algorithm in #1875 and added the missing unit tests to cover the edge cases.

#### Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/ect)

#### Checklist:
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Common/Securities/Option/OptionFilterUniverse.cs [hunk 1]
```diff
@@ -220,27 +220,68 @@ public OptionFilterUniverse Strikes(int minStrike, int maxStrike)
             _isDynamic = true;
 
             // find the current price in the list of strikes
+            var exactPriceFound = true;
             var index = _uniqueStrikes.BinarySearch(_underlying.Price);
+
+            // Return value of BinarySearch (from MSDN):
+            // The zero-based index of item in the sorted List<T>, if item is found;
+            // otherwise, a negative number that is the bitwise complement of the index of the next element that is larger than item
```
Comment (mchandschuh) [inline Common/Securities/Option/OptionFilterUniverse.cs]: :+1:  -- these are GREAT comments to include, thank you.

File: Common/Securities/Option/OptionFilterUniverse.cs [hunk 2]
```diff
@@ -220,27 +220,68 @@ public OptionFilterUniverse Strikes(int minStrike, int maxStrike)
...
+                    _allSymbols = Enumerable.Empty<Symbol>();
+                    return this;
+                }
+
                 index = ~index;
 
                 if (index >= _uniqueStrikes.Count)
                 {
-                    // price out of range: should never happen, return empty
+                    // should never happen, return empty
```
Comment (mchandschuh) [inline Common/Securities/Option/OptionFilterUniverse.cs]: anytime I see 'should never happen' I think we should throw an exception or _minimally_ log as much state as possible
